stan/math/rev/core/var.hpp

```diff
@@ -25,36 +23,42 @@ static void grad(vari* vi);
  * an arena-based memory manager scoped to a single gradient
  * calculation.
  *
- * A var is constructed with a double and used like any
- * other scalar.  Arithmetical functions like negation, addition,
+ * A var is constructed with a type `T` and used like any
+ * other scalar. Arithmetical functions like negation, addition,
  * and subtraction, as well as a range of mathematical functions
  * like exponentiation and powers are overridden to operate on
  * var values objects.
+ * @tparam T An Floating point type.
  */
-class var {
+template <typename T, typename = void>
+class var_value {};
+
+template <typename T>
+class var_value<T, require_floating_point_t<T>> {
  public:
-  // FIXME: doc what this is for
-  using Scalar = double;
+  using value_type = std::decay_t<T>;        // Numeric type in vari_value.
+  using vari_type = vari_value<value_type>;  // Type of underlying vari impl.
+  using vari_pointer = vari_type*;  // pointer type for underlying vari.
```

Comment:
I'd prefer we just use `vari_type*` when we need a pointer. Clearer what that means (there's nothing fancy like a `shared_ptr` or anything happening under the hood)

Reply:
Yeah I agree. I was thinking about including member types like [`std::vector`](https://en.cppreference.com/w/cpp/container/vector) has but using `vari_type*` is also fine by me

Reply:
Can still delete this type